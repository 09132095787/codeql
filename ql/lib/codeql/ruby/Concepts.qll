/**
 * Provides abstract classes representing generic concepts such as file system
 * access or system command execution, for which individual framework libraries
 * provide concrete subclasses.
 */

private import codeql.ruby.AST
private import codeql.ruby.CFG
private import codeql.ruby.DataFlow
private import codeql.ruby.Frameworks
private import codeql.ruby.dataflow.RemoteFlowSources
private import codeql.ruby.ApiGraphs

/**
 * A data-flow node that executes SQL statements.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `SqlExecution::Range` instead.
 */
class SqlExecution extends DataFlow::Node instanceof SqlExecution::Range {
  /** Gets the argument that specifies the SQL statements to be executed. */
  DataFlow::Node getSql() { result = super.getSql() }
}

/** Provides a class for modeling new SQL execution APIs. */
module SqlExecution {
  /**
   * A data-flow node that executes SQL statements.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `SqlExecution` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Gets the argument that specifies the SQL statements to be executed. */
    abstract DataFlow::Node getSql();
  }
}

/**
 * A data-flow node that escapes meta-characters, which could be used to prevent
 * injection attacks.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `Escaping::Range` instead.
 */
class Escaping extends DataFlow::Node instanceof Escaping::Range {
  Escaping() {
    // escapes that don't have _both_ input/output defined are not valid
    exists(super.getAnInput()) and
    exists(super.getOutput())
  }

  /** Gets an input that will be escaped. */
  DataFlow::Node getAnInput() { result = super.getAnInput() }

  /** Gets the output that contains the escaped data. */
  DataFlow::Node getOutput() { result = super.getOutput() }

  /**
   * Gets the context that this function escapes for, such as `html`, or `url`.
   */
  string getKind() { result = super.getKind() }
}

/** Provides a class for modeling new escaping APIs. */
module Escaping {
  /**
   * A data-flow node that escapes meta-characters, which could be used to prevent
   * injection attacks.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `Escaping` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Gets an input that will be escaped. */
    abstract DataFlow::Node getAnInput();

    /** Gets the output that contains the escaped data. */
    abstract DataFlow::Node getOutput();

    /**
     * Gets the context that this function escapes for.
     *
     * While kinds are represented as strings, this should not be relied upon. Use the
     * predicates in  the `Escaping` module, such as `getHtmlKind`.
     */
    abstract string getKind();
  }

  /** Gets the escape-kind for escaping a string so it can safely be included in HTML. */
  string getHtmlKind() { result = "html" }
}

/**
 * An escape of a string so it can be safely included in
 * the body of an HTML element, for example, replacing `{}` in
 * `<p>{}</p>`.
 */
class HtmlEscaping extends Escaping {
  HtmlEscaping() { super.getKind() = Escaping::getHtmlKind() }
}

/** Provides classes for modeling HTTP-related APIs. */
module HTTP {
  /** Provides classes for modeling HTTP servers. */
  module Server {
    /**
     * A data-flow node that sets up a route on a server.
     *
     * Extend this class to refine existing API models. If you want to model new APIs,
     * extend `RouteSetup::Range` instead.
     */
    class RouteSetup extends DataFlow::Node instanceof RouteSetup::Range {
      /** Gets the URL pattern for this route, if it can be statically determined. */
      string getUrlPattern() { result = super.getUrlPattern() }

      /**
       * Gets a function that will handle incoming requests for this route, if any.
       *
       * NOTE: This will be modified in the near future to have a `RequestHandler` result, instead of a `Method`.
       */
      Method getARequestHandler() { result = super.getARequestHandler() }

      /**
       * Gets a parameter that will receive parts of the url when handling incoming
       * requests for this route, if any. These automatically become a `RemoteFlowSource`.
       */
      Parameter getARoutedParameter() { result = super.getARoutedParameter() }

      /** Gets a string that identifies the framework used for this route setup. */
      string getFramework() { result = super.getFramework() }
    }

    /** Provides a class for modeling new HTTP routing APIs. */
    module RouteSetup {
      /**
       * A data-flow node that sets up a route on a server.
       *
       * Extend this class to model new APIs. If you want to refine existing API models,
       * extend `RouteSetup` instead.
       */
      abstract class Range extends DataFlow::Node {
        /** Gets the argument used to set the URL pattern. */
        abstract DataFlow::Node getUrlPatternArg();

        /** Gets the URL pattern for this route, if it can be statically determined. */
        string getUrlPattern() {
          exists(CfgNodes::ExprNodes::StringlikeLiteralCfgNode strNode |
            this.getUrlPatternArg().getALocalSource() = DataFlow::exprNode(strNode) and
            result = strNode.getExpr().getValueText()
          )
        }

        /**
         * Gets a function that will handle incoming requests for this route, if any.
         *
         * NOTE: This will be modified in the near future to have a `RequestHandler` result, instead of a `Method`.
         */
        abstract Method getARequestHandler();

        /**
         * Gets a parameter that will receive parts of the url when handling incoming
         * requests for this route, if any. These automatically become a `RemoteFlowSource`.
         */
        abstract Parameter getARoutedParameter();

        /** Gets a string that identifies the framework used for this route setup. */
        abstract string getFramework();
      }
    }

    /**
     * A function that will handle incoming HTTP requests.
     *
     * Extend this class to refine existing API models. If you want to model new APIs,
     * extend `RequestHandler::Range` instead.
     */
    class RequestHandler extends Method instanceof RequestHandler::Range {
      /**
       * Gets a parameter that could receive parts of the url when handling incoming
       * requests, if any. These automatically become a `RemoteFlowSource`.
       */
      Parameter getARoutedParameter() { result = super.getARoutedParameter() }

      /** Gets a string that identifies the framework used for this route setup. */
      string getFramework() { result = super.getFramework() }
    }

    /** Provides a class for modeling new HTTP request handlers. */
    module RequestHandler {
      /**
       * A function that will handle incoming HTTP requests.
       *
       * Extend this class to model new APIs. If you want to refine existing API models,
       * extend `RequestHandler` instead.
       *
       * Only extend this class if you can't provide a `RouteSetup`, since we handle that case automatically.
       */
      abstract class Range extends Method {
        /**
         * Gets a parameter that could receive parts of the url when handling incoming
         * requests, if any. These automatically become a `RemoteFlowSource`.
         */
        abstract Parameter getARoutedParameter();

        /** Gets a string that identifies the framework used for this request handler. */
        abstract string getFramework();
      }
    }

    private class RequestHandlerFromRouteSetup extends RequestHandler::Range {
      RouteSetup rs;

      RequestHandlerFromRouteSetup() { this = rs.getARequestHandler() }

      override Parameter getARoutedParameter() {
        result = rs.getARoutedParameter() and
        result = this.getAParameter()
      }

      override string getFramework() { result = rs.getFramework() }
    }

    /** A parameter that will receive parts of the url when handling an incoming request. */
    private class RoutedParameter extends RemoteFlowSource::Range, DataFlow::ParameterNode {
      RequestHandler handler;

      RoutedParameter() { this.getParameter() = handler.getARoutedParameter() }

      override string getSourceType() { result = handler.getFramework() + " RoutedParameter" }
    }

    /**
     * A data-flow node that creates a HTTP response on a server.
     *
     * Note: we don't require that this response must be sent to a client (a kind of
     * "if a tree falls in a forest and nobody hears it" situation).
     *
     * Extend this class to refine existing API models. If you want to model new APIs,
     * extend `HttpResponse::Range` instead.
     */
    class HttpResponse extends DataFlow::Node instanceof HttpResponse::Range {
      /** Gets the data-flow node that specifies the body of this HTTP response. */
      DataFlow::Node getBody() { result = super.getBody() }

      /** Gets the mimetype of this HTTP response, if it can be statically determined. */
      string getMimetype() { result = super.getMimetype() }
    }

    /** Provides a class for modeling new HTTP response APIs. */
    module HttpResponse {
      /**
       * A data-flow node that creates a HTTP response on a server.
       *
       * Note: we don't require that this response must be sent to a client (a kind of
       * "if a tree falls in a forest and nobody hears it" situation).
       *
       * Extend this class to model new APIs. If you want to refine existing API models,
       * extend `HttpResponse` instead.
       */
      abstract class Range extends DataFlow::Node {
        /** Gets the data-flow node that specifies the body of this HTTP response. */
        abstract DataFlow::Node getBody();

        /** Gets the data-flow node that specifies the content-type/mimetype of this HTTP response, if any. */
        abstract DataFlow::Node getMimetypeOrContentTypeArg();

        /** Gets the default mimetype that should be used if `getMimetypeOrContentTypeArg` has no results. */
        abstract string getMimetypeDefault();

        /** Gets the mimetype of this HTTP response, if it can be statically determined. */
        string getMimetype() {
          exists(CfgNodes::ExprNodes::StringlikeLiteralCfgNode strNode |
            this.getMimetypeOrContentTypeArg().getALocalSource() = DataFlow::exprNode(strNode) and
            result = strNode.getExpr().getValueText().splitAt(";", 0)
          )
          or
          not exists(this.getMimetypeOrContentTypeArg()) and
          result = this.getMimetypeDefault()
        }
      }
    }

    /**
     * A data-flow node that creates a HTTP redirect response on a server.
     *
     * Note: we don't require that this redirect must be sent to a client (a kind of
     * "if a tree falls in a forest and nobody hears it" situation).
     *
     * Extend this class to refine existing API models. If you want to model new APIs,
     * extend `HttpRedirectResponse::Range` instead.
     */
    class HttpRedirectResponse extends HttpResponse instanceof HttpRedirectResponse::Range {
      /** Gets the data-flow node that specifies the location of this HTTP redirect response. */
      DataFlow::Node getRedirectLocation() { result = super.getRedirectLocation() }
    }

    /** Provides a class for modeling new HTTP redirect response APIs. */
    module HttpRedirectResponse {
      /**
       * A data-flow node that creates a HTTP redirect response on a server.
       *
       * Note: we don't require that this redirect must be sent to a client (a kind of
       * "if a tree falls in a forest and nobody hears it" situation).
       *
       * Extend this class to model new APIs. If you want to refine existing API models,
       * extend `HttpResponse` instead.
       */
      abstract class Range extends HTTP::Server::HttpResponse::Range {
        /** Gets the data-flow node that specifies the location of this HTTP redirect response. */
        abstract DataFlow::Node getRedirectLocation();
      }
    }
  }
}

/**
 * A data flow node that executes an operating system command,
 * for instance by spawning a new process.
 */
class SystemCommandExecution extends DataFlow::Node {
  SystemCommandExecution::Range range;

  SystemCommandExecution() { this = range }

  /** Holds if a shell interprets `arg`. */
  predicate isShellInterpreted(DataFlow::Node arg) { range.isShellInterpreted(arg) }

  /** Gets an argument to this execution that specifies the command or an argument to it. */
  DataFlow::Node getAnArgument() { result = range.getAnArgument() }
}

module SystemCommandExecution {
  /**
   * A data flow node that executes an operating system command, for instance by spawning a new
   * process.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `SystemCommandExecution` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Gets an argument to this execution that specifies the command or an argument to it. */
    abstract DataFlow::Node getAnArgument();

    /** Holds if a shell interprets `arg`. */
    predicate isShellInterpreted(DataFlow::Node arg) { none() }
  }
}
