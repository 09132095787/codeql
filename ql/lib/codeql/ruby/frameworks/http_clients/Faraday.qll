private import ruby
private import codeql.ruby.Concepts
private import codeql.ruby.ApiGraphs

/**
 * A call that makes an HTTP request using `Faraday`.
 * ```ruby
 * # one-off request
 * Faraday.get("http://example.com").body
 *
 * # connection re-use
 * connection = Faraday.new("http://example.com")
 * connection.get("/").body
 * ```
 */
class FaradayHttpRequest extends HTTP::Client::Request::Range {
  DataFlow::Node requestUse;
  API::Node requestNode;
  API::Node connectionNode;

  FaradayHttpRequest() {
    connectionNode =
      [
        // one-off requests
        API::getTopLevelMember("Faraday"),
        // connection re-use
        API::getTopLevelMember("Faraday").getInstance()
      ] and
    requestNode =
      connectionNode.getReturn(["get", "head", "delete", "post", "put", "patch", "trace"]) and
    requestUse = requestNode.getAnImmediateUse() and
    this = requestUse.asExpr().getExpr()
  }

  override DataFlow::Node getResponseBody() { result = requestNode.getAMethodCall("body") }

  override predicate disablesCertificateValidation(DataFlow::Node disablingNode) {
    // `Faraday::new` takes an options hash as its second argument, and we're
    // looking for
    // `{ ssl: { verify: false } }`
    // or
    // `{ ssl: { verify_mode: OpenSSL::SSL::VERIFY_NONE } }`
    exists(DataFlow::Node arg, int i |
      i > 0 and arg = connectionNode.getAUse().(DataFlow::CallNode).getArgument(i)
    |
      // Either passed as an individual key:value argument, e.g.:
      // Faraday.new(..., ssl: {...})
      isSslOptionsPairDisablingValidation(arg.asExpr().getExpr()) and
      disablingNode = arg
      or
      // Or as a single hash argument, e.g.:
      // Faraday.new(..., { ssl: {...} })
      exists(DataFlow::LocalSourceNode optionsNode, Pair p |
        p = optionsNode.asExpr().getExpr().(HashLiteral).getAKeyValuePair() and
        isSslOptionsPairDisablingValidation(p) and
        optionsNode.flowsTo(arg) and
        disablingNode.asExpr().getExpr() = p
      )
    )
  }

  override string getFramework() { result = "Faraday" }
}

/**
 * Holds if the pair `p` contains the key `:ssl` for which the value is a hash
 * containing either `verify: false` or
 * `verify_mode: OpenSSL::SSL::VERIFY_NONE`.
 */
private predicate isSslOptionsPairDisablingValidation(Pair p) {
  exists(DataFlow::Node key, DataFlow::Node value |
    key.asExpr().getExpr() = p.getKey() and value.asExpr().getExpr() = p.getValue()
  |
    exists(DataFlow::LocalSourceNode literal |
      literal.asExpr().getExpr().(SymbolLiteral).getValueText() = "ssl" and
      literal.flowsTo(key)
    ) and
    (isHashWithVerifyFalse(value) or isHashWithVerifyModeNone(value))
  )
}

/**
 * Holds if `node` represents a hash containing the key-value pair
 * `verify: false`.
 */
private predicate isHashWithVerifyFalse(DataFlow::Node node) {
  exists(DataFlow::LocalSourceNode hash |
    isVerifyFalsePair(hash.asExpr().getExpr().(HashLiteral).getAKeyValuePair()) and
    hash.flowsTo(node)
  )
}

/**
 * Holds if `node` represents a hash containing the key-value pair
 * `verify_mode: OpenSSL::SSL::VERIFY_NONE`.
 */
private predicate isHashWithVerifyModeNone(DataFlow::Node node) {
  exists(DataFlow::LocalSourceNode hash |
    isVerifyModeNonePair(hash.asExpr().getExpr().(HashLiteral).getAKeyValuePair()) and
    hash.flowsTo(node)
  )
}

/**
 * Holds if the pair `p` has the key `:verify_mode` and the value
 * `OpenSSL::SSL::VERIFY_NONE`.
 */
private predicate isVerifyModeNonePair(Pair p) {
  exists(DataFlow::Node key, DataFlow::Node value |
    key.asExpr().getExpr() = p.getKey() and value.asExpr().getExpr() = p.getValue()
  |
    exists(DataFlow::LocalSourceNode literal |
      literal.asExpr().getExpr().(SymbolLiteral).getValueText() = "verify_mode" and
      literal.flowsTo(key)
    ) and
    value = API::getTopLevelMember("OpenSSL").getMember("SSL").getMember("VERIFY_NONE").getAUse()
  )
}

/**
 * Holds if the pair `p` has the key `:verify` and the value `false`.
 */
private predicate isVerifyFalsePair(Pair p) {
  exists(DataFlow::Node key, DataFlow::Node value |
    key.asExpr().getExpr() = p.getKey() and value.asExpr().getExpr() = p.getValue()
  |
    exists(DataFlow::LocalSourceNode literal |
      literal.asExpr().getExpr().(SymbolLiteral).getValueText() = "verify" and
      literal.flowsTo(key)
    ) and
    isFalsey(value)
  )
}

/** Holds if `node` contains `0` or `false`. */
private predicate isFalsey(DataFlow::Node node) {
  exists(DataFlow::LocalSourceNode literal |
    (
      literal.asExpr().getExpr().(BooleanLiteral).isFalse() or
      literal.asExpr().getExpr().(IntegerLiteral).getValue() = 0
    ) and
    literal.flowsTo(node)
  )
}
