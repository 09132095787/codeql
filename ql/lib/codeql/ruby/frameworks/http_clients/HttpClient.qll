private import ruby
private import codeql.ruby.Concepts
private import codeql.ruby.ApiGraphs

/**
 * A call that makes an HTTP request using `HTTPClient`.
 * ```ruby
 * HTTPClient.get("http://example.com").body
 * HTTPClient.get_content("http://example.com")
 * ```
 */
class HttpClientRequest extends HTTP::Client::Request::Range {
  DataFlow::Node request;
  DataFlow::CallNode responseBody;

  HttpClientRequest() {
    exists(API::Node requestNode, string method |
      request = requestNode.getAnImmediateUse() and
      method in [
          "get", "head", "delete", "options", "post", "put", "trace", "get_content", "post_content"
        ]
    |
      requestNode = API::getTopLevelMember("HTTPClient").getReturn(method) and
      (
        // The `get_content` and `post_content` methods return the response body as a string.
        // The other methods return a `HTTPClient::Message` object which has various methods
        // that return the response body.
        method in ["get_content", "post_content"] and responseBody = request
        or
        not method in ["get_content", "put_content"] and
        responseBody = requestNode.getAMethodCall(["body", "http_body", "content", "dump"])
      ) and
      this = request.asExpr().getExpr()
    )
  }

  override DataFlow::Node getResponseBody() { result = responseBody }

  override string getFramework() { result = "HTTPClient" }
}
