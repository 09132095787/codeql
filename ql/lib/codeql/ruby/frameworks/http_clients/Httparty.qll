private import ruby
private import codeql.ruby.Concepts
private import codeql.ruby.ApiGraphs

/**
 * A call that makes an HTTP request using `HTTParty`.
 * ```ruby
 * # one-off request - returns the response body
 * HTTParty.get("http://example.com")
 *
 * # TODO: module inclusion
 * class MyClass
 *  include HTTParty
 * end
 *
 * MyClass.new("http://example.com")
 * ```
 */
class HttpartyRequest extends HTTP::Client::Request::Range {
  DataFlow::Node request;
  DataFlow::CallNode responseBody;

  HttpartyRequest() {
    exists(API::Node requestNode | request = requestNode.getAnImmediateUse() |
      requestNode =
        API::getTopLevelMember("HTTParty")
            .getReturn(["get", "head", "delete", "options", "post", "put", "patch"]) and
      (
        // If HTTParty can recognise the response type, it will parse and return it
        // directly from the request call. Otherwise, it will return a `HTTParty::Response`
        // object that has a `#body` method.
        // So if there's a call to `#body` on the response, treat that as the response body.
        exists(DataFlow::Node r | r = requestNode.getAMethodCall("body") | responseBody = r)
        or
        // Otherwise, treat the response as the response body.
        not exists(DataFlow::Node r | r = requestNode.getAMethodCall("body")) and
        responseBody = request
      ) and
      this = request.asExpr().getExpr()
    )
  }

  override DataFlow::Node getResponseBody() { result = responseBody }

  override string getFramework() { result = "HTTParty" }
}
