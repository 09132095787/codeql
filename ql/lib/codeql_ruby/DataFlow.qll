/**
 * Provides classes for performing local (intra-procedural) and
 * global (inter-procedural) data flow analyses.
 */
module DataFlow {
  import codeql_ruby.dataflow.internal.DataFlowImpl
}
