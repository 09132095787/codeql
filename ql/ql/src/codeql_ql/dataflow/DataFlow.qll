private import codeql_ql.ast.Ast
private import internal.NodesInternal

/**
 * An expression or variable in a formula, including some additional nodes
 * that are not part of the AST.
 */
class Node extends TNode {
  string toString() { none() } // overridden in subclasses
  Location getLocation() { none() } // overridden in subclasses

  /**
   * Gets the underlying `Expr` or `VarDef` node, if this is an `AstNodeNode`.
   */
  AstNode asAstNode() { astNode(result) = this }

  /**
   * Gets the predicate containing this data-flow node.
   *
   * All data-flow nodes belong in exactly one predicate.
   * TODO: select clauses
   */
  Predicate getEnclosingPredicate() { none() } // overridden in subclasses
}

/**
 * A data-flow node based an `Expr` or `VarDef` AST node.
 */
class AstNodeNode extends Node, MkAstNodeNode {
  private AstNode ast;

  AstNodeNode() { this = MkAstNodeNode(ast) }

  override string toString() {
    result = ast.toString()
  }

  override Location getLocation() {
    result = ast.getLocation()
  }

  /** Gets the AST node. */
  AstNode getAstNode() {
    result = ast
  }

  override Predicate getEnclosingPredicate() {
    result = ast.getEnclosingPredicate()
  }
}

/**
 * Gets the data-flow node correspoinding to the given AST node.
 */
pragma[inline]
Node astNode(AstNode node) {
  result = MkAstNodeNode(node)
}

/**
 * A data-flow node representing a variable within a specific scope.
 */
class ScopedVariableNode extends Node, MkScopedVariable {
  private VarDef var;
  private AstNode scope;

  ScopedVariableNode() { this = MkScopedVariable(var, scope) }

  override string toString() {
    result = "Variable '" + var.getName() + "' scoped to " + scope.getLocation().getStartLine() + ":" + scope.getLocation().getStartColumn()
  }

  override Location getLocation() {
    result = scope.getLocation()
  }

  /** Gets the variable being refined to a specific scope. */
  VarDef getVariable() {
    result = var
  }

  /** Gets the scope to which this variable has been refined. */
  AstNode getScope() {
    result = scope
  }

  override Predicate getEnclosingPredicate() {
    result = var.getEnclosingPredicate()
  }
}

/**
 * Gets the data-flow node corresponding to `var` restricted to `scope`.
 */
pragma[inline]
Node scopedVariable(VarDef var, AstNode scope) {
  result = MkScopedVariable(var, scope)
}

/**
 * A data-flow node representing `this` within a class predicate, charpred, or newtype branch.
 */
class ThisNode extends Node, MkThisNode {
  private Predicate pred;

  ThisNode() { this = MkThisNode(pred) }

  override string toString() {
    result = "'this' in " + pred.getName()
  }

  override Location getLocation() {
    result = pred.getLocation()
  }

  /** Gets the class predicate, charpred, or newtype branch whose 'this' parameter is represented by this node. */
  Predicate getPredicate() {
    result = pred
  }

  override Predicate getEnclosingPredicate() {
    result = pred
  }
}

/**
 * Gets the data-flow node representing `this` within the given class predicate, charpred, or newtype branch.
 */
pragma[inline]
Node thisNode(Predicate pred) {
  result = MkThisNode(pred)
}

/**
 * A data-flow node representing `result` within a predicate that has a result.
 */
class ResultNode extends Node, MkResultNode {
  private Predicate pred;

  ResultNode() { this = MkResultNode(pred) }

  override string toString() {
    result = "'result' in " + pred.getName()
  }

  override Location getLocation() {
    result = pred.getLocation()
  }

  /** Gets the predicate whose 'result' parameter is represented by this node. */
  Predicate getPredicate() {
    result = pred
  }

  override Predicate getEnclosingPredicate() {
    result = pred
  }
}

/**
 * Gets the data-flow node representing `result` within the given predicate.
 */
pragma[inline]
Node resultNode(Predicate pred) {
  result = MkResultNode(pred)
}

/**
 * A data-flow node representing the view of a field in the enclosing class, as seen
 * from a charpred or class predicate.
 */
class FieldNode extends Node, MkFieldNode {
  private Predicate pred;
  private FieldDecl fieldDecl;

  FieldNode() { this = MkFieldNode(pred, fieldDecl) }

  /** Gets the member predicate or charpred for which this node represents access to the field. */
  Predicate getPredicate() {
    result = pred
  }

  FieldDecl getFieldDeclaration() {
    result = fieldDecl
  }

  string getFieldName() {
    result = fieldDecl.getName()
  }

  override string toString() {
    result = "'"  + this.getFieldName() + "' in " + pred.getName()
  }

  override Location getLocation() {
    result = pred.getLocation()
  }

  override Predicate getEnclosingPredicate() {
    result = pred
  }
}

/**
 * Gets the data-flow node representing the given predicate's view of the given field
 * in the enclosing class.
 */
pragma[inline]
Node fieldNode(Predicate pred, FieldDecl fieldDecl) {
  result = MkFieldNode(pred, fieldDecl)
}
