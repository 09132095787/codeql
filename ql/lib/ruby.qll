import codeql_ruby.AST
