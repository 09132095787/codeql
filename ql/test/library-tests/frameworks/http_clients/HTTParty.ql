import codeql.ruby.frameworks.http_clients.HTTParty
import codeql.ruby.DataFlow

query DataFlow::Node httpartyRequests(HTTPartyRequest e) { result = e.getResponseBody() }
