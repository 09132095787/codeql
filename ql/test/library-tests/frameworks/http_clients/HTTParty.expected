| HTTParty.rb:5:1:5:35 | call to get | HTTParty.rb:5:1:5:35 | call to get |
| HTTParty.rb:7:1:7:55 | call to post | HTTParty.rb:7:1:7:55 | call to post |
| HTTParty.rb:9:1:9:54 | call to put | HTTParty.rb:9:1:9:54 | call to put |
| HTTParty.rb:11:1:11:56 | call to patch | HTTParty.rb:11:1:11:56 | call to patch |
| HTTParty.rb:15:9:15:46 | call to delete | HTTParty.rb:16:1:16:10 | call to body |
| HTTParty.rb:18:9:18:44 | call to head | HTTParty.rb:19:1:19:10 | call to body |
| HTTParty.rb:21:9:21:47 | call to options | HTTParty.rb:22:1:22:10 | call to body |
